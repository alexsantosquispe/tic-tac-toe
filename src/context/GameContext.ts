import { createContext } from 'react';
import type { SquareValueTypes } from '../types';

export interface GameContextTypes {
  data: SquareValueTypes[];
  currentPlayer: 'X' | 'O';
  winner: SquareValueTypes | null;
  combHighLight: number[];
  checkSquare: (index: number) => void;
  toggleCurrentPlayer: () => void;
}

const GameContext = createContext<GameContextTypes | null>(null);

export default GameContext;
