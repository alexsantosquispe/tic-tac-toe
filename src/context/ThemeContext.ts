import { createContext } from 'react';

export const THEME_TYPES = {
  SYSTEM: 'system',
  LIGHT: 'light',
  DARK: 'dark'
} as const;

export type ThemeType = (typeof THEME_TYPES)[keyof typeof THEME_TYPES];

interface ThemeContextType {
  theme: ThemeType;
  setTheme: (theme: ThemeType) => void;
}

export const ThemeContext = createContext<ThemeContextType | null>(null);
