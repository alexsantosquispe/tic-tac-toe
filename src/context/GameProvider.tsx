import { useState, type ReactNode } from 'react';
import { DEFAULT_DATA } from '../constants';
import type { SquareValueTypes } from '../types';
import { getWinnerResult } from '../utils';
import GameContext from './GameContext';

interface GameProviderProps {
  children: ReactNode;
}

const GameProvider = ({ children }: GameProviderProps) => {
  const [currentPlayer, setCurrentPlayer] = useState<'X' | 'O'>('X');
  const [data, setData] = useState<SquareValueTypes[]>(DEFAULT_DATA);
  const [winner, setWinner] = useState<SquareValueTypes | null>(null);
  const [combHighLight, setCombHighLight] = useState<number[]>([]);

  const toggleCurrentPlayer = () => {
    setCurrentPlayer((prevPlayer) => (prevPlayer === 'X' ? 'O' : 'X'));
  };

  const checkSquare = (index: number) => {
    if (data[index] !== '' || winner) return;
    const newData = [...data];
    newData[index] = currentPlayer;
    setData(newData);
    const result = getWinnerResult(newData);
    if (!result.winner) {
      toggleCurrentPlayer();
    }
    setWinner(result.winner);
    setCombHighLight(result.combHighLight);
  };

  const resetGame = () => {
    setData(DEFAULT_DATA);
    setWinner(null);
    setCombHighLight([]);
    setCurrentPlayer('X');
  };

  return (
    <GameContext.Provider
      value={{
        data,
        currentPlayer,
        winner,
        combHighLight,
        checkSquare,
        toggleCurrentPlayer,
        resetGame
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export default GameProvider;
