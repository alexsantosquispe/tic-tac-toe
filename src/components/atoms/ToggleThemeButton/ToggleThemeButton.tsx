import cn from 'clsx';
import type { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';
import { THEME_TYPES, type ThemeType } from '../../../context/ThemeContext';
import { useTheme } from '../../../hooks/useTheme';
import { MoonIcon } from '../../../icons/MoonIcon';
import { SunIcon } from '../../../icons/SunIcon';
import { SystemIcon } from '../../../icons/SystemIcon';

const THEME_OPTIONS = [
  { id: THEME_TYPES.SYSTEM, icon: <SystemIcon /> },
  { id: THEME_TYPES.LIGHT, icon: <SunIcon /> },
  { id: THEME_TYPES.DARK, icon: <MoonIcon /> }
];

interface ThemeOptionProps {
  id: ThemeType;
  icon: ReactNode;
  isSelected: boolean;
  onSelectOption: (theme: ThemeType) => void;
}

export const ThemeOption = ({
  id,
  icon,
  isSelected = false,
  onSelectOption
}: ThemeOptionProps) => {
  return (
    <button
      aria-label={`${id} button`}
      className={twMerge(
        'rounded-full border-neutral-200 p-0.5 text-neutral-600 hover:cursor-pointer dark:border-white/20',
        cn({ 'border bg-rose-600 text-white': isSelected })
      )}
      onClick={() => onSelectOption(id)}
    >
      {icon}
    </button>
  );
};

export const ToggleThemeButton = () => {
  const { theme, setTheme } = useTheme();

  return (
    <div className="flex items-center gap-1.5 rounded-full border border-neutral-200 p-1 dark:border-white/20">
      {THEME_OPTIONS.map((option) => {
        const isSelected = option.id === theme;
        return (
          <ThemeOption
            key={option.id}
            {...option}
            isSelected={isSelected}
            onSelectOption={setTheme}
          />
        );
      })}
    </div>
  );
};
