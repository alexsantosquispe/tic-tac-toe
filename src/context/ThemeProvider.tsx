import { useEffect, useState, type ReactNode } from 'react';
import {
  applyTheme,
  getStoredTheme,
  getSystemTheme,
  setStoredTheme
} from '../utils/theme.utils';
import { THEME_TYPES, ThemeContext, type ThemeType } from './ThemeContext';

interface ThemeProviderProps {
  children: ReactNode;
}

const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const initialTheme = getStoredTheme();
  const [currentTheme, setCurrentTheme] = useState<ThemeType>(
    initialTheme ?? THEME_TYPES.SYSTEM
  );

  useEffect(() => {
    const theme =
      currentTheme === THEME_TYPES.SYSTEM ? getSystemTheme() : currentTheme;
    applyTheme(theme);
    setStoredTheme(currentTheme);
  }, [currentTheme]);

  useEffect(() => {
    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    const handleChange = (e: MediaQueryListEvent) => {
      if (getStoredTheme() === THEME_TYPES.SYSTEM) {
        applyTheme(e.matches ? THEME_TYPES.DARK : THEME_TYPES.LIGHT);
      }
    };
    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  return (
    <ThemeContext.Provider
      value={{ theme: currentTheme, setTheme: setCurrentTheme }}
    >
      {children}
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;
