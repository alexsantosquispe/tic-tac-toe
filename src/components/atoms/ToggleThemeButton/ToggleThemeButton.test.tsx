import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import { THEME_TYPES } from '../../../context/ThemeContext';
import ThemeProvider from '../../../context/ThemeProvider';
import { ToggleThemeButton } from './ToggleThemeButton';

describe('ToggleThemButton', () => {
  describe('styles', () => {
    it('should render the component correctly', () => {
      const component = render(
        <ThemeProvider>
          <ToggleThemeButton />
        </ThemeProvider>
      );

      expect(component).toMatchSnapshot();

      expect(
        screen.getByRole('button', { name: `${THEME_TYPES.SYSTEM} button` })
      );
      expect(
        screen.getByRole('button', { name: `${THEME_TYPES.DARK} button` })
      );
      expect(
        screen.getByRole('button', { name: `${THEME_TYPES.LIGHT} button` })
      );
    });
  });

  describe.only('behavior', () => {
    beforeEach(() => {
      render(
        <ThemeProvider>
          <ToggleThemeButton />
        </ThemeProvider>
      );
    });

    it('should change the theme', () => {
      const selectedButtonStyles = 'text-white';
      const systemButton = screen.getByRole('button', {
        name: `${THEME_TYPES.SYSTEM} button`
      });
      const lightButton = screen.getByRole('button', {
        name: `${THEME_TYPES.LIGHT} button`
      });
      const darkButton = screen.getByRole('button', {
        name: `${THEME_TYPES.DARK} button`
      });
      const indicator = screen.getByTestId('selected-indicator');

      expect(systemButton).toHaveClass(selectedButtonStyles);
      expect(lightButton).not.toHaveClass(selectedButtonStyles);
      expect(darkButton).not.toHaveClass(selectedButtonStyles);
      expect(indicator).toHaveClass('translate-x-0');

      fireEvent.click(lightButton);
      expect(systemButton).not.toHaveClass(selectedButtonStyles);
      expect(lightButton).toHaveClass(selectedButtonStyles);
      expect(darkButton).not.toHaveClass(selectedButtonStyles);
      expect(indicator).toHaveClass('translate-x-[2rem]');

      fireEvent.click(darkButton);
      expect(systemButton).not.toHaveClass(selectedButtonStyles);
      expect(lightButton).not.toHaveClass(selectedButtonStyles);
      expect(darkButton).toHaveClass(selectedButtonStyles);
      expect(indicator).toHaveClass('translate-x-[4rem]');
    });
  });
});
