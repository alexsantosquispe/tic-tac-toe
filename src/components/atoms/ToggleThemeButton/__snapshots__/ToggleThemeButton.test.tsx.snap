// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ToggleThemButton styles should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
      >
        <div
          class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-black dark:bg-white"
          data-testid="selected-indicator"
        />
        <button
          aria-label="system button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer text-white dark:text-black"
        >
          <svg
            class="size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 5a2 2 0 0 1 2 2v8.526a2 2 0 0 0 .212.897l1.068 2.127a1 1 0 0 1-.9 1.45H3.62a1 1 0 0 1-.9-1.45l1.068-2.127A2 2 0 0 0 4 15.526V7a2 2 0 0 1 2-2z"
            />
            <path
              d="M20.054 15.987H3.946"
            />
          </svg>
        </button>
        <button
          aria-label="light button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <svg
            class="size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="4"
            />
            <path
              d="M12 2v2"
            />
            <path
              d="M12 20v2"
            />
            <path
              d="m4.93 4.93 1.41 1.41"
            />
            <path
              d="m17.66 17.66 1.41 1.41"
            />
            <path
              d="M2 12h2"
            />
            <path
              d="M20 12h2"
            />
            <path
              d="m6.34 17.66-1.41 1.41"
            />
            <path
              d="m19.07 4.93-1.41 1.41"
            />
          </svg>
        </button>
        <button
          aria-label="dark button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <svg
            class="size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.985 12.486a9 9 0 1 1-9.473-9.472c.405-.022.617.46.402.803a6 6 0 0 0 8.268 8.268c.344-.215.825-.004.803.401"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
    >
      <div
        class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-black dark:bg-white"
        data-testid="selected-indicator"
      />
      <button
        aria-label="system button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer text-white dark:text-black"
      >
        <svg
          class="size-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 5a2 2 0 0 1 2 2v8.526a2 2 0 0 0 .212.897l1.068 2.127a1 1 0 0 1-.9 1.45H3.62a1 1 0 0 1-.9-1.45l1.068-2.127A2 2 0 0 0 4 15.526V7a2 2 0 0 1 2-2z"
          />
          <path
            d="M20.054 15.987H3.946"
          />
        </svg>
      </button>
      <button
        aria-label="light button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
      >
        <svg
          class="size-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="4"
          />
          <path
            d="M12 2v2"
          />
          <path
            d="M12 20v2"
          />
          <path
            d="m4.93 4.93 1.41 1.41"
          />
          <path
            d="m17.66 17.66 1.41 1.41"
          />
          <path
            d="M2 12h2"
          />
          <path
            d="M20 12h2"
          />
          <path
            d="m6.34 17.66-1.41 1.41"
          />
          <path
            d="m19.07 4.93-1.41 1.41"
          />
        </svg>
      </button>
      <button
        aria-label="dark button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
      >
        <svg
          class="size-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20.985 12.486a9 9 0 1 1-9.473-9.472c.405-.022.617.46.402.803a6 6 0 0 0 8.268 8.268c.344-.215.825-.004.803.401"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
